### ISSUE TITLE
allow expression with custom function in Model object

### ISSUE THREAD

**Issue by Jerry-Ma** [lmfit/lmfit-py#304]
Hi, thank you for the awesome package. 

I just found that it would be useful if we can define custom function for defining derived parameters in Model object. Here is my thought:

Currently, the way we use model interface is as follows:
    1. define a `Model` object using a model function
    2. call `fit` method of the model object to obtain best fit parameters

What step 2 does is that:
    1. creates a  `ModelResult` object, which is a `Minimizer`
    2. call the fitting method and update the parameters

If we can have access to the ModelRresult object before the fitting, so we can update the ast attribute to include any custom expression.

**Comment by newville** [comment]
@Jerry-Ma  Is it possible to do what you want by first creating a Model, then creating and modifying the Parameters, as described with `Model.make_params()` and `Model.set_param_hint()`?   See https://lmfit.github.io/lmfit-py/model.html#initializing-values-with-model-make-params
for reference.

**Comment by Jerry-Ma** [comment]
@newville Thank you for advise. I was able to achieve what I said by modifying the `Parameters` object, not after the creation though, as I found the constructor of `Parameter` class takes an optional `asteval` argument, which should be a `asteval.Interpreter` object, and the latter support custom function through `Interpreter.symtable` attribute.

Obviously, I had to change the source code to accommodate the solution.

**Comment by newville** [comment]
@Jerry-Ma Is there something that  Model.make_params() and Model.set_param_hint() were not able to do that required direct access to the Interpreter symbol table?   I would think that you could create an instance of Parameters, then add/alter its symbol table.   

You raised this as an issue, but I don't understand what that issue would be.  It sounds like you found a solution that involved changing the source code.  I don't see why it is obvious that you had to change the source code to do what you wanted to do.  Can you explain why that was necessary?   Is this change something that should be added to the code?

**Comment by Jerry-Ma** [comment]
@newville Thank you for the comment. The reason why I didn't go through the route you suggested was that I wasn't able to find the proper interface for adding user defined symtable entry after the creation of parameters class.

Yes, in principle, I could do `params = Model.make_params()` and `params._asteval.symtable.update(mysymtable)`, but it looks hacky, as by convention, user should not make use of the `_*` methods.

The modification to the code I made was just adding a keyword argument to the Model creator, which is a dict of the user defined functions. When `make_params()` is called the dict is updated to the symtable associated with the created `Parameters` object.

I would suggest close this issue or if you think its appropriate, I can make a PR.

**Comment by newville** [comment]
@Jerry-Ma  Since you were clearly instructed to read Contributing.md when you submitted this Issue, I will assume that you are responsible and polite enough to have followed those instructions.   If you haven't read that, do so now.

OK, so now that you have definitely read Contributing.md, your Issue is definitely saying that there is a problem with the code.     But you have not provided an example or even a clear description of the problem that you raised, even though I asked for clarification twice.   So, again, what is the problem?   Why couldn't it be solved as I suggested?  What would you make a PR for?

**Comment by Jerry-Ma** [comment]
I actually just found the existence of the Contributing.md. I feel now that it seems not appropriate to open an issue here after read it.

I thought I had provided the picture there but clearly it's so scattered around that it's hard to be perceived.

What is the problem?
===
    The `Model` and/or `Model.fit` do not provide an interface for adding custom functions to be used for defining constrained/derived parameters. The lack of this interface makes `Model.make_params` less useful, because if you do things like:
    model = Model(modelfunc)    
    model.set_param_hint('fo')
    model.set_param_hint('ob', expr='ar(fo)')
    model.fit()
It would fail as function `ar` is not defined.

Why couldn't it be solved as I suggested?
===
    I could be solved as you suggested, which I suppose the following untested code:
    model = Model(modelfunc)    
    model.set_param_hint('fo')
    model.set_param_hint('ob', expr='ar(fo)')
    par  = model.make_params()
    par._asteval.symtable.update(mycustomsymtable)
    model.fit(params=par)  # using supplied param
 
What would you make a PR for? 
===
    As I pointed out, the above solution is not optimal, as by convention (as for as I know), use should prevent as much as possible accessing the `_*` methods/attributes. 
    It I were to make a PR, it would utilize the fact that `asteval.Interpreter` has `symtable` attribute and I can add my custom function there, and then instantiate the `Parameters` (this action is at the very beginning of `Model.make_params` method) class by constructor `Parameters(asteval=myInterpreter)`. This could be enabled by adding an optional keyword argument either in `Model.__init__` or `Model.make_params`.

Summary, in any case, the problem I was facing was solved. And next time I would first post on the mail list for questions alike.

**Comment by newville** [comment]
@Jerry-Ma  You say:

> I actually just found the existence of the Contributing.md. 

Can you help me out by explaining how you didn't see this earlier?  I believe you must have pointed a browser at https://github.com/lmfit/lmfit-py/issues/new  to create a new Issue.  You must have done that all consciously and deliberately to actively raise an Issue.    When you browse there it says as clear as it possibly can 

```
 Please review the guidelines for contributing to this repository.
```

We cannot control the formatting of that web page to, say, make the font larger.

You must have typed into the Title bar right below this message.  You must have typed into the text box below that.   How can it possibly be that you did not even see that request?  

Alternatively, you *could* have read the docs, where the main page clearly (really, go look again) says where to go with questions.  Apparently you did not look there either.

I am continually dismayed at people using github Issues for non-Issues and do not understand why this keeps happening.   So, really, I do not mean to pick on you, but I simply do not understand how this can possibly be happening.    Why did you raise an Issue?  What was your goal and motivation?  Yes, I would really like answers to those questions.

I think it may be necessary to simply close Issues as off-topic unless the submitter is a developer or has been vetted through the mailing list.  Anyone else have any thoughts?  Like, even if you think I have sn irrational hatred of using github Issues for questions, wouldn't it be fair to say that if someone does not see or cannot be bothered to read the docs or Contributing.md that they will not be submitting a useful or helpful Issue?   I think the correlation of "didn't bother to read Contributing.md" and "very poorly defined Issue" is very high.

As for the "Issue" you wanted to talk about, that seems like a pretty small problem to me.  In short, one *can* access the asteval Interpreter.

**Comment by Jerry-Ma** [comment]
> So, really, I do not mean to pick on you, but I simply do not understand how this can possibly be happening. Why did you raise an Issue? What was your goal and motivation? Yes, I would really like answers to those questions.
From my personal experience here, I would say the reason is that I was not clearly aware of the purpose of `issue tracker`. You may argue that it is clearly notified when opening a new issue, for reading the `contribute.md`, however, what happened to me was that I wasn't clear about that asking questions is called `contribute`. Of course this wouldn't happen if I have had known the mailing-list.
Of course, this is not a general answer but I am sure that the diversity of maintenance styles of different repos contributes to what you saw, because I have done similar things in other places, and I mistakenly thought it applies to larger extent.
In summary, I would say as long as there is a large bulk of users who used to only `use` but some day they are trying to `contribute` or contribute,  incidents like this would be expected to happen.
I suggest close the issue as off-topic.

**Comment by newville** [comment]
@Jerry-Ma Thanks for the reply -- I do really appreciate that and find it helpful.   

I'm aware that others view Github Issues as a viable, general-purpose support forum,  and it seems to have a lower barrier of entry for many people.  Personally, I find it lacking in a number of ways.   We've tried to do everything we can to point people toward the mailing list instead of using Issues, but the success rate appears to be about 50%.

For the actual issue, manipulating `Parameter._asteval` seems completely reasonable to me, with the single underscore suggesting "here be dragons, beware!" rather than "don't use".   It might be reasonable to add a `Parameter.add_symbol(name, value)` to allow adding a user-defined function or other data into the symbol table.    Either approach seems like an advanced use for people who know what they're doing, so I don't have a strong preference for which approach is preferred / sanctioned.

**Comment by Tillsten** [comment]
@newville https://github.com/blog/2111-issue-and-pull-request-templates

**Comment by newville** [comment]
@tillsten that is great news!  Such  templates will be very useful.    I'm very glad to see github.com responding to this problem.
